File: app/src/main/java/com/grassehh/app/controller/SampleController.java
package com.grassehh.app.controller;

import io.micrometer.tracing.Tracer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.reactive.function.client.WebClient;
import org.springframework.web.util.UriComponentsBuilder;
import reactor.core.publisher.Mono;

import java.util.UUID;

import static org.springframework.http.HttpStatus.OK;

@RestController
@RequestMapping("/test")
public class SampleController {
    private WebClient webClient;
    private Tracer tracer;

    private static Logger logger = LoggerFactory.getLogger(SampleController.class);

    public SampleController(WebClient webClient, Tracer tracer) {
        this.webClient = webClient;
        this.tracer = tracer;
    }

    @GetMapping
    @ResponseStatus(OK)
    @PreAuthorize("hasRole('USER')")
    public Mono<Void> test() {
        tracer.createBaggageInScope("myBaggageController", UUID.randomUUID().toString());
        logger.debug("Log from inside the controller. This should contain the baggage called myBaggageController");
        return webClient.get().uri(UriComponentsBuilder.newInstance().scheme("https").host("google.com").build().toUri()).retrieve().bodyToMono(String.class).then(Mono.empty());
    }
}

File: app/src/main/kotlin/com/grassehh/app/controller/SampleController.kt
package com.grassehh.app.controller

import com.grassehh.app.service.ProxiedService
import io.micrometer.core.instrument.kotlin.asContextElement
import io.micrometer.observation.ObservationRegistry
import io.micrometer.tracing.Tracer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import mu.KotlinLogging.logger
import org.springframework.http.HttpStatus.OK
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitExchange
import java.util.*

@RestController
class SampleController(
    private val webClient: WebClient,
    private val tracer: Tracer,
    private val observationRegistry: ObservationRegistry,
    private val proxiedService: ProxiedService
) {
    companion object {
        private val logger = logger {}
    }

    @GetMapping("simple")
    @ResponseStatus(OK)
    @PreAuthorize("hasRole('USER')")
    suspend fun simple() {
        tracer.createBaggageInScope("myBaggageController", UUID.randomUUID().toString())
        logger.debug { "CONTROLLER LOG" }
        webClient.get()
            .uri("https://www.google.fr")
            .awaitExchange { }
    }

    @GetMapping("suspend")
    @ResponseStatus(OK)
    @PreAuthorize("hasRole('USER')")
    suspend fun suspend() {
        tracer.createBaggageInScope("myBaggageController", UUID.randomUUID().toString())
        delay(500)
        logger.debug { "CONTROLLER LOG" }
        webClient.get()
            .uri("https://www.google.fr")
            .awaitExchange { }
    }

    @GetMapping("coroutine")
    @ResponseStatus(OK)
    @PreAuthorize("hasRole('USER')")
    suspend fun coroutine() {
        tracer.createBaggageInScope("myBaggageController", UUID.randomUUID().toString())
        logger.debug { "CONTROLLER LOG BEFORE COROUTINE" }
        CoroutineScope(SupervisorJob()).launch(observationRegistry.asContextElement()) {
            logger.debug { "CONTROLLER LOG INSIDE COROUTINE" }
            webClient.get()
                .uri("https://www.google.fr")
                .awaitExchange { }
        }
    }

    @GetMapping("aop")
    @ResponseStatus(OK)
    @PreAuthorize("hasRole('USER')")
    suspend fun aop() {
        tracer.createBaggageInScope("myBaggageController", UUID.randomUUID().toString())
        logger.debug { "CONTROLLER LOG BEFORE COROUTINE" }
        CoroutineScope(SupervisorJob()).launch(observationRegistry.asContextElement()) {
            logger.debug { "CONTROLLER LOG INSIDE COROUTINE" }
            proxiedService.doSomethingSuspend()
        }
    }
}
